contrib/CONTRIBUTING.md

```diff
@@ -0,0 +1,232 @@
+- Use `black` to format your Python code; it ensures readability and consistency.
+- Write concise Git commit messages; summarize changes in ~50 characters.
+- Follow these six commit rules:
+  - Atomic Commits: Focus on one task or fix per commit.
+  - Subject and Body Separation: Use a blank line to separate the subject from the body.
+  - Subject Line Length: Keep it under 50 characters for readability.
+  - Imperative Mood: Write subject line as if giving a command or instruction.
+  - Body Text Width: Wrap text manually at 72 characters.
+  - Body Content: Explain what changed and why, not how.
+- Make use of your commit messages to simplify project understanding and maintenance.
+
+### Code Contribution General Guidelines
+
+> Review the Bittensor [style guide](./STYLE.md) and [development workflow](./DEVELOPMENT_WORKFLOW.md) before contributing. 
+
+If you're looking to contribute to Bittensor but unsure where to start, please join our community [discord](https://discord.gg/bittensor), a developer-friendly Bittensor town square. Start with [#development](https://discord.com/channels/799672011265015819/799678806159392768) and [#bounties](https://discord.com/channels/799672011265015819/1095684873810890883) to see what issues are currently posted. For a greater understanding of Bittensor's usage and development, check the [Bittensor Documentation](https://bittensor.com/docs).
+
+
+
+#### Pull Request Philosophy
+
+Patchsets and enhancements should always be focused. For example, a pull request could add a feature, fix a bug, or refactor code; but not a mixture. Please also avoid super pull requests which attempt to do too much, are overly large, or overly complex as this makes review difficult. This helps maintain a separation of concerns and reduces review cognitive load.
+
+In general, all pull requests must:
+
+  - Have a clear use case, fix a demonstrable bug or serve the greater good of the project (for example refactoring for modularisation);
+  - Be well peer-reviewed;
+  - Have unit tests and functional tests, where appropriate;
+  - Follow code style guidelines;
+  - Not break the existing test suite;
+  - Where bugs are fixed, where possible, there should be unit tests demonstrating the bug and also proving the fix. This helps prevent regression.
+  - Change relevant comments and documentation when behaviour of code changes.
+
+#### Pull Request Process
+Please follow these steps to have your contribution considered by the maintainers:
+
+*Before* creating the PR:
+1. Read the [development workflow](./DEVELOPMENT_WORKFLOW.md) defined for this repository in order to agree on the ways of working.
+1. While coding, please, add tests relevant to the fixed bug or new feature. Follow the [testing guide](./TESTING.md)
+1. To create the PR follow all instructions in [the template](https://github.com/opentensor/bittensor/blob/master/.github/PULL_REQUEST_TEMPLATE/pull_request_template.md)
```

Comment:
A [pull request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) should be added which should force people to give a summary of the changes and reference the related issues, a list of changes, how testing was done, and a checklist regarding final steps such as documentation updates, compatibility checks, and that the changelog has been updated.

The changelog in general should be part of the changes - a list of the titles of the merged PRs with links to the PRs (which generally have no body) is not a changelog.

Reply:
How is this different from the above comment about ISSUE_TEMPLATES?

Should we unify these two things into a set of PULL_REQUEST TEMPLATEs?  Seem to be conflicting/confusing to have both.

Reply:
Ditto @eduardogr if you have ability to do so:

Bug-Fixes:
```yaml
name: Bug Fix Contribution
description: Use this template when contributing a bug fix.
labels: [bug, pull request]

body:
  - type: markdown
    attributes:
      value: |
        ### Requirements for Contributing a Bug Fix

        * Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
        * The pull request must only fix an existing bug. To contribute other changes, you must use a different template. You can see all templates at <https://github.com/atom/.github/tree/master/.github/PULL_REQUEST_TEMPLATE>.
        * The pull request must update the test suite to demonstrate the changed functionality. For guidance, please see <https://flight-manual.atom.io/hacking-atom/sections/writing-specs/>.
        * After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see <https://github.com/atom/.github/tree/master/CONTRIBUTING.md#pull-requests>.

  - type: input
    id: bug
    attributes:
      label: Identify the Bug
      description: Link to the issue describing the bug that you're fixing.
    validations:
      required: true

  - type: textarea
    id: change
    attributes:
      label: Description of the Change
      description: We must be able to understand the design of your change from this description.
    validations:
      required: true

  - type: textarea
    id: alternate
    attributes:
      label: Alternate Designs
      description: Explain what other alternates were considered and why the proposed version was selected.
    validations:
      required: false

  - type: textarea
    id: drawbacks
    attributes:
      label: Possible Drawbacks
      description: What are the possible side-effects or negative impacts of the code change?
    validations:
      required: false

  - type: textarea
    id: verification
    attributes:
      label: Verification Process
      description: What process did you follow to verify that the change has not introduced any regressions?
    validations:
      required: true

  - type: input
    id: release-notes
    attributes:
      label: Release Notes
      description: Please describe the changes in a single line that explains this improvement in terms that a user can understand.
    validations:
      required: true

```


Performance:
```yaml

name: Performance Improvement Contribution
description: Use this template when contributing a performance improvement.
labels: [performance, pull request]

body:
  - type: markdown
    attributes:
      value: |
        ### Requirements for Contributing a Performance Improvement

        * Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
        * The pull request must only affect performance of existing functionality. To contribute other changes, you must use a different template. You can see all templates at <https://github.com/atom/.github/tree/master/.github/PULL_REQUEST_TEMPLATE>.
        * After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see <https://github.com/atom/.github/tree/master/CONTRIBUTING.md#pull-requests>.

  - type: textarea
    id: change
    attributes:
      label: Description of the Change
      description: We must be able to understand the design of your change from this description.
    validations:
      required: true

  - type: textarea
    id: benefits
    attributes:
      label: Quantitative Performance Benefits
      description: Describe the exact performance improvement observed.
    validations:
      required: true

  - type: textarea
    id: drawbacks
    attributes:
      label: Possible Drawbacks
      description: What are the possible side-effects or negative impacts of the code change?
    validations:
      required: false

  - type: textarea
    id: verification
    attributes:
      label: Verification Process
      description: What process did you follow to verify that your change has the desired effects?
    validations:
      required: true

  - type: input
    id: applicable-issues
    attributes:
      label: Applicable Issues
      description: Enter any applicable Issues here.
    validations:
      required: false

  - type: input
    id: release-notes
    attributes:
      label: Release Notes
      description: Please describe the changes in a single line that explains this improvement in terms that a user can understand.
    validations:
      required: true

```

Features:
```yaml
name: Feature Contribution
description: Use this template when adding, changing, or removing a feature.
labels: [feature, pull request]

body:
  - type: markdown
    attributes:
      value: |
        ### Requirements for Adding, Changing, or Removing a Feature

        * Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
        * The pull request must contribute a change that has been endorsed by the maintainer team. See details in the template below.
        * The pull request must update the test suite to exercise the updated functionality. For guidance, please see <https://flight-manual.atom.io/hacking-atom/sections/writing-specs/>.
        * After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see <https://github.com/atom/.github/tree/master/CONTRIBUTING.md#pull-requests>.

  - type: input
    id: issue-or-rfc
    attributes:
      label: Issue or RFC Endorsed by Atom's Maintainers
      description: Link to the issue or RFC that your change relates to.
    validations:
      required: true

  - type: textarea
    id: change
    attributes:
      label: Description of the Change
      description: We must be able to understand the design of your change from this description.
    validations:
      required: true

  - type: textarea
    id: alternate
    attributes:
      label: Alternate Designs
      description: Explain what other alternates were considered and why the proposed version was selected.
    validations:
      required: false

  - type: textarea
    id: drawbacks
    attributes:
      label: Possible Drawbacks
      description: What are the possible side-effects or negative impacts of the code change?
    validations:
      required: false

  - type: textarea
    id: verification
    attributes:
      label: Verification Process
      description: What process did you follow to verify that your change has the desired effects?
    validations:
      required: true

  - type: input
    id: release-notes
    attributes:
      label: Release Notes
      description: Please describe the changes in a single line that explains this improvement in terms that a user can understand.
    validations:
      required: true

```

Reply:
As mentioned offline, while there are _beta_ _issue forms_, there are no _pull request forms_, only _pull request templates_. The pull request template has to stay as markdown, but it will still be prefilled when someone tries to open a PR.

Reply:
Reverted.  Thanks!